Automattic/woocommerce-payments
# Direct Checkout Optimization: Preload Cart Session Data

Fixes 2481-gh-Automattic/woopay.

#### Changes proposed in this Pull Request









This PR optimizes the WooPay Direct Checkout flow by preloading the WooPay session data. It also handles cases where the cart is updated. It does this for both the classic and wcBlocks cart.

This PR also addresses the following:
* Prevents the "Buy with WooPay" button from re-rendering when clicked.
* Adds the mechanism that prevents the WooPay Direct Checkout flow from being applied, from the frontend, when the feature is supposed to be disabled.
* Ensures the WooPay Direct Checkout flow is only applied to the cart page.
* Ensures the WooPay session data is not preloaded from the product page.
* Ensures that the WooPay Direct Checkout flow is still applied when the classic cart is updated.

#### Additional Notes

Special attention should be paid to the usage of the following wcBlocks hooks:
* `experimental__woocommerce_blocks-cart-add-item`.
* `experimental__woocommerce_blocks-cart-set-item-quantity`.
* `experimental__woocommerce_blocks-cart-remove-item`.

Although these hooks are prefixed with "experimental", they are deemed to be stable and there are plans to graduate them (with the "experimental" prefix) in the near future. For more details, see the following link:
* p1708601365661539/1708555986.110799-slack-C02TS23QJ1X

#### Testing instructions





**Regression Test: Ensure Purchasing a Product With 1 Stock Does Not Fail** (see #7256 for details)
* Ensure the WooPay Direct Checkout flow is enabled.
* Create an item with only 1 stock and run the following test scenarios (you will have to update the stock back to 1 after each test):
    * Use the WooPay button from the cart page and pay through the WooPay checkout.
    * Navigate to the cart page before going to the Merchant checkout and pay through the Merchant checkout (make sure not to type a valid WooPay address in the email field).
    * Navigate to the cart page before navigating to the Merchant checkout and pay through the WooPay Button.
    * Navigate to the cart page before navigating to the Merchant checkout and pay using the WooPay email input.
    * Navigate directly to the Merchant checkout and pay through the Merchant checkout (make sure not to type a valid woopay address in the email field).
    * Navigate directly to the Merchant checkout and pay through the WooPay button.
    * Navigate directly to the Merchant checkout and pay through the WooPay email input.
* Repeat these tests for both the classic cart/checkout and the wcBlocks cart/checkout.

**Test: Ensure WooPay Direct Checkout Flow Is Only Applied in the Cart Page**
* Ensure the WooPay Direct Checkout flow is enabled.
* As a shopper, navigate to the shop and add an item to your cart.
* Navigate to the cart page and ensure the `woopay-direct-checkout.js` script is fetched by inspecting the Network tab in the browser's developer tools.
* Navigate to a product and ensure the `woopay-direct-checkout.js` script is NOT fetched.
* Navigate to the classic/wcBlocks checkout page and ensure the `woopay-direct-checkout.js` script is NOT fetched.

**Test: Ensure WooPay Session Data Is Preloaded, On Page Load**
* Ensure the following conditions are true:
  * WooPay Direct Checkout flow is enabled.
  * Third-party cookies are enabled.
  * A valid WooPay user is authenticated on WooPay.
* As a shopper, add multiple items to your cart with any given quantity. 
* Navigate to the cart page and click the "Proceed to Checkout" button.
* Ensure you are navigated to WooPay and order summary contains the correct quantities of your items.
* Ensure you are able to successfully checkout through WooPay by clicking the "Place order" button.
* Repeat this for both the classic and wcBlocks cart pages.

**Test: Ensure WooPay Session Data Is Preloaded, On Cart Update**
* Ensure the following conditions are true:
  * WooPay Direct Checkout flow is enabled.
  * Third-party cookies are enabled.
  * A valid WooPay user is authenticated on WooPay.
* As a shopper, add multiple items to your cart with any given quantity. 
* Navigate to the cart page.
* Update the cart page by adding products, removing products, and updating product quantities.
* Click the "Proceed to Checkout" button.
* Ensure you are navigated to WooPay and order summary contains the correct quantities of your items.
* Ensure you are able to successfully checkout through WooPay by clicking the "Place order" button.
* Repeat this for both the classic and wcBlocks cart pages.

-------------------


**Post merge**
